File: gdx-pay-iosmoe-apple/native/GdxPayStoreKit2Bridge/Sources/GdxPayStoreKit2Bridge/GDXStoreKit2Bridge.swift
import Foundation
import StoreKit

@objc(GDXStoreKit2Bridge)
@available(iOS 15.0, macOS 12.0, *)
public final class GDXStoreKit2Bridge: NSObject {
    @objc public static let shared = GDXStoreKit2Bridge()

    private var productsById: [String: Product] = [:]
    private var transactionUpdatesTask: Task<Void, Never>?

    @objc public func canMakePayments() -> Bool {
        AppStore.canMakePayments
    }

    @objc(fetchProductsWithIdentifiers:completion:)
    public func fetchProducts(
        withIdentifiers identifiers: [String],
        completion: @escaping ([[String: Any]]?, NSError?) -> Void
    ) {
        Task {
            do {
                let products = try await Product.products(for: identifiers)
                productsById = Dictionary(uniqueKeysWithValues: products.map { ($0.id, $0) })

                var mappedProducts: [[String: Any]] = []
                for product in products {
                    mappedProducts.append(await map(product))
                }

                completion(mappedProducts, nil)
            } catch {
                completion(nil, error as NSError)
            }
        }
    }

    @objc(purchaseWithIdentifier:completion:)
    public func purchase(
        withIdentifier identifier: String,
        completion: @escaping ([String: Any]?, NSError?) -> Void
    ) {
        Task {
            guard let product = productsById[identifier] else {
                completion(nil, bridgeError(code: 1, message: "Product not loaded: \(identifier)"))
                return
            }

            do {
                let result = try await product.purchase()

                switch result {
                case .success(let verificationResult):
                    let transaction = try checked(verificationResult)
                    await transaction.finish()
                    completion(map(transaction), nil)

                case .userCancelled:
                    completion(["cancelled": true, "productID": identifier], nil)

                case .pending:
                    completion(["pending": true, "productID": identifier], nil)

                @unknown default:
                    completion(nil, bridgeError(code: 2, message: "Unknown purchase result for \(identifier)"))
                }
            } catch {
                completion(nil, error as NSError)
            }
        }
    }

    @objc(fetchCurrentEntitlementsWithCompletion:)
    public func fetchCurrentEntitlements(
        completion: @escaping ([[String: Any]]?, NSError?) -> Void
    ) {
        Task {
            do {
                var transactions: [[String: Any]] = []

                for await result in Transaction.currentEntitlements {
                    let transaction = try checked(result)
                    transactions.append(map(transaction))
                }

                completion(transactions, nil)
            } catch {
                completion(nil, error as NSError)
            }
        }
    }

    @objc(restorePurchasesWithCompletion:)
    public func restorePurchases(
        completion: @escaping ([[String: Any]]?, NSError?) -> Void
    ) {
        Task {
            do {
                try await AppStore.sync()
                fetchCurrentEntitlements(completion: completion)
            } catch {
                completion(nil, error as NSError)
            }
        }
    }

    @objc(startObservingTransactionsWithCompletion:)
    public func startObservingTransactions(
        completion: @escaping ([String: Any]?, NSError?) -> Void
    ) {
        stopObservingTransactions()

        transactionUpdatesTask = Task {
            for await result in Transaction.updates {
                do {
                    let transaction = try checked(result)
                    await transaction.finish()
                    completion(map(transaction), nil)
                } catch {
                    completion(nil, error as NSError)
                }
            }
        }
    }

    @objc public func stopObservingTransactions() {
        transactionUpdatesTask?.cancel()
        transactionUpdatesTask = nil
    }

    private func checked<T>(_ result: VerificationResult<T>) throws -> T {
        switch result {
        case .verified(let value):
            return value
        case .unverified(_, let error):
            throw error
        }
    }

    private func map(_ product: Product) async -> [String: Any] {
        var mapped: [String: Any] = [
            "id": product.id,
            "displayName": product.displayName,
            "description": product.description,
            "displayPrice": product.displayPrice,
            "currencyCode": product.priceFormatStyle.currencyCode,
            "price": NSDecimalNumber(decimal: product.price)
        ]

        if let subscription = product.subscription,
           let introductoryOffer = subscription.introductoryOffer {
            mapped["eligibleForIntroOffer"] = await subscription.isEligibleForIntroOffer
            mapped["introPeriodValue"] = introductoryOffer.period.value
            mapped["introPeriodUnit"] = map(introductoryOffer.period.unit)
            mapped["introPrice"] = NSDecimalNumber(decimal: introductoryOffer.price)
        }

        return mapped
    }

    private func map(_ transaction: Transaction) -> [String: Any] {
        [
            "productID": transaction.productID,
            "originalID": String(transaction.originalID),
            "purchaseDate": transaction.purchaseDate.timeIntervalSince1970,
            "jsonRepresentation": String(data: transaction.jsonRepresentation, encoding: .utf8) ?? "",
            "jsonRepresentationBase64": transaction.jsonRepresentation.base64EncodedString()
        ]
    }

    private func map(_ unit: Product.SubscriptionPeriod.Unit) -> String {
        switch unit {
        case .day:
            return "day"
        case .week:
            return "week"
        case .month:
            return "month"
        case .year:
            return "year"
        @unknown default:
            return "unknown"
        }
    }

    private func bridgeError(code: Int, message: String) -> NSError {
        NSError(
            domain: "com.badlogic.gdx.pay.iosmoe.storekit2",
            code: code,
            userInfo: [NSLocalizedDescriptionKey: message]
        )
    }
}

File: gdx-pay-iosmoe-apple/src/main/java/com/badlogic/gdx/pay/ios/apple/bindings/GDXStoreKit2Bridge.java
package com.badlogic.gdx.pay.ios.apple.bindings;

import apple.NSObject;
import apple.foundation.NSArray;
import apple.foundation.NSDictionary;
import apple.foundation.NSError;

import org.moe.natj.general.NatJ;
import org.moe.natj.general.Pointer;
import org.moe.natj.general.ann.Library;
import org.moe.natj.general.ann.Runtime;
import org.moe.natj.objc.ObjCRuntime;
import org.moe.natj.objc.ann.ObjCBlock;
import org.moe.natj.objc.ann.ObjCClassBinding;
import org.moe.natj.objc.ann.ObjCClassName;
import org.moe.natj.objc.ann.Selector;

@Library("GdxPayStoreKit2Bridge")
@Runtime(ObjCRuntime.class)
@ObjCClassBinding
@ObjCClassName("GDXStoreKit2Bridge")
public class GDXStoreKit2Bridge extends NSObject {
    static {
        NatJ.register();
    }

    protected GDXStoreKit2Bridge(Pointer peer) {
        super(peer);
    }

    @Selector("shared")
    public static native GDXStoreKit2Bridge shared();

    @Selector("canMakePayments")
    public native boolean canMakePayments();

    @Selector("fetchProductsWithIdentifiers:completion:")
    public native void fetchProductsWithIdentifiersCompletion(
            NSArray<String> identifiers,
            @ObjCBlock(name = "call_fetchProductsWithIdentifiersCompletion")
            Block_fetchProductsWithIdentifiersCompletion completion);

    @Selector("purchaseWithIdentifier:completion:")
    public native void purchaseWithIdentifierCompletion(
            String identifier,
            @ObjCBlock(name = "call_purchaseWithIdentifierCompletion")
            Block_purchaseWithIdentifierCompletion completion);

    @Selector("fetchCurrentEntitlementsWithCompletion:")
    public native void fetchCurrentEntitlementsWithCompletion(
            @ObjCBlock(name = "call_fetchCurrentEntitlementsWithCompletion")
            Block_fetchCurrentEntitlementsWithCompletion completion);

    @Selector("restorePurchasesWithCompletion:")
    public native void restorePurchasesWithCompletion(
            @ObjCBlock(name = "call_restorePurchasesWithCompletion")
            Block_restorePurchasesWithCompletion completion);

    @Selector("startObservingTransactionsWithCompletion:")
    public native void startObservingTransactionsWithCompletion(
            @ObjCBlock(name = "call_startObservingTransactionsWithCompletion")
            Block_startObservingTransactionsWithCompletion completion);

    @Selector("stopObservingTransactions")
    public native void stopObservingTransactions();

    public interface Block_fetchProductsWithIdentifiersCompletion {
        void call_fetchProductsWithIdentifiersCompletion(NSArray<NSDictionary<String, Object>> products, NSError error);
    }

    public interface Block_purchaseWithIdentifierCompletion {
        void call_purchaseWithIdentifierCompletion(NSDictionary<String, Object> transaction, NSError error);
    }

    public interface Block_fetchCurrentEntitlementsWithCompletion {
        void call_fetchCurrentEntitlementsWithCompletion(NSArray<NSDictionary<String, Object>> transactions, NSError error);
    }

    public interface Block_restorePurchasesWithCompletion {
        void call_restorePurchasesWithCompletion(NSArray<NSDictionary<String, Object>> transactions, NSError error);
    }

    public interface Block_startObservingTransactionsWithCompletion {
        void call_startObservingTransactionsWithCompletion(NSDictionary<String, Object> transaction, NSError error);
    }
}
